src/tribler/core/utilities/slow_coro_detection.py

```diff
@@ -0,0 +1,136 @@
+from asyncio import Handle, Task
+from threading import Lock, Thread
+
+from typing import Optional
+
+from tribler.core.utilities.main_thread_stack_tracking import get_main_thread_stack
+
+from tribler.core.utilities.main_thread_stack_tracking import main_stack_tracking_is_activated
+
+# pylint: disable=protected-access
+
+logger = logging.getLogger(__name__)
+
+SLOW_CORO_DURATION_THRESHOLD = 1.0
+WATCHING_THREAD_INTERVAL = 1.0
+
+
+class DebugInfo:
+    def __init__(self):
+        self.handle = None
+        self.start_time = None
+
+
+current = DebugInfo()
+lock = Lock()
+
+
+class SlowCoroWatchingThread(Thread):
+    """
+    A thread that detects and reports slow coroutines.
+    """
+    def run(self):
+        prev_reported_handle = None  # to detect in the loop when we have the second report of the same slow coroutine
+        while True:
+            time.sleep(WATCHING_THREAD_INTERVAL)
+            with lock:
+                handle, start_time = current.handle, current.start_time
+            if handle is not None:
+                duration = time.time() - start_time
+                if duration > SLOW_CORO_DURATION_THRESHOLD:
+                    report_freeze(current.handle, duration, first_report=(prev_reported_handle is not handle))
+                    prev_reported_handle = handle
+                else:
+                    prev_reported_handle = None
+
+
+def patched_handle_run(self: Handle):
+    """
+    Remembers the current asyncio handle object and its starting time globally, so it becomes possible
+    to access it from the separate thread and detect slow coroutines.
+    """
+    start_time = time.time()
+    with lock:
+        current.handle, current.start_time = self, start_time
+    try:
+        original_handle_run(self)
+    finally:
+        with lock:
+            assert current.handle is self, "asyncio loop should never run one handle inside another one"
+            current.handle = current.start_time = None
+
+        duration = time.time() - start_time
+        if duration > SLOW_CORO_DURATION_THRESHOLD:
+            # The coroutine step is finished successfully (without freezing), but the execution time was too long
+            report_long_duration(self, duration)
+
+        self = None  # Needed to break cycles when an exception occurs (copied from the original Handle._run method)
+
+
+original_handle_run = Handle._run
+
+
+def report_long_duration(handle: Handle, duration: float):
+    info_str = format_info(handle)
+    logger.error(f'Slow coroutine step execution (duration={duration:.3f} seconds): {info_str}')
+
+
+def report_freeze(handle: Handle, duration: float, first_report: bool):
+    if not first_report:
+        info_str = format_info(handle, include_stack=False)
+        logger.error(f"Still executing {info_str}")
+        return
+
+    info_str = format_info(handle, include_stack=True)
+    logger.error(f'Slow coroutine is occupying the loop for {duration:.3f} seconds already: {info_str}')
```

Comment:
This function would be easier to read if the blocks were rearranged, with the first report coming first, followed by all others.

```suggestion
def report_freeze(handle: Handle, duration: float, first_report: bool):
    if first_report:
        info_str = format_info(handle, include_stack=True)
        logger.error(f'Slow coroutine is occupying the loop for {duration:.3f} seconds already: {info_str}')
        return

    info_str = format_info(handle, include_stack=False)
    logger.error(f"Still executing {info_str}")
```

Reply:
Changed